src/Functions/GregorianDate.h

```
@@ -329,18 +330,30 @@ namespace DB
     }
 
     template <typename YearT>
-    OrdinalDate<YearT>::OrdinalDate(is_integer auto mjd)
+    template <is_integer DayT>
+    OrdinalDate<YearT>::OrdinalDate(DayT modified_julian_day)
     {
-        const auto a         = mjd + 678575;
+        /// This function supports day number from -678941 to 2973119 (which represent 0000-01-01 and 9999-12-31 respectively).
+
+        if constexpr (is_signed_v<DayT> && std::numeric_limits<DayT>::lowest() < -678941)
+            if (modified_julian_day < -678941)
+                modified_julian_day = -678941;
+
+        if constexpr (std::numeric_limits<DayT>::max() > 2973119)
+            if (modified_julian_day > 2973119)
+                modified_julian_day = 2973119;
```

This changed the behaviour a bit which was caught by the failing test.
Previously it would fail to parse datetime but we clamp.
Maybe it's better to throw an error here because it's more clear?

---

Yes, let's throw an exception here.